src/rust/engine/graph/src/lib.rs

```diff
@@ -878,19 +881,28 @@ impl<N: Node> Graph<N> {
     run_token: RunToken,
     result: Option<Result<N::Item, N::Error>>,
   ) {
-    let (entry, entry_id, dep_generations) = {
+    let (entry, complete_as_dirty, dep_generations) = {
       let inner = self.inner.lock();
+      let mut complete_as_dirty = false;
       // Get the Generations of all dependencies of the Node. We can trust that these have not changed
       // since we began executing, as long as we are not currently marked dirty (see the method doc).
       let dep_generations = inner
         .pg
         .neighbors_directed(entry_id, Direction::Outgoing)
         .filter_map(|dep_id| inner.entry_for_id(dep_id))
-        .map(Entry::generation)
+        .map(|entry| {
+          // If a dependency is uncacheable or currently dirty, this Node should complete as dirty,
+          // independent of matching Generation values. This is to allow for the behaviour that an
+          // uncacheable Node should always have dirty dependents, transitively.
+          if !entry.node().cacheable(context) || !entry.is_clean(context) {
+            complete_as_dirty = true;
```

Comment:
`complete_as_dirty` is ignored if `!cacheable` - should this maybe be a tri-state enum we can match on, rather than a function call which we'll repeat and a bool?

Reply:
> rather than a function call which we'll repeat and a bool?

We don't repeat the function call... the calls to `cacheable(..)` here are for the dependencies, while the call inside `Entry::complete` is for the node itself. Given that, I don't think the enum makes sense.

Reply:
On re-read, I think I see what you are suggesting... can try that.

Reply:
So, this isn't straightforward: the `complete_as_dirty` boolean comes into play for most of the states of the `Entry::complete(.., result)` parameter, which means that replacing those two parameters with an enum would require a five state enum I think. I tried it and ran into a roadblock.

Reply:
Interesting... Do you know what the names of those 5 enum variants would be? I'm curious as to whether that enum would be more clear (even if it's more verbose), given that my read of this code was that there would be 3 enum variants, it seems like I mis-understood the behaviour as it's currently written...

Reply:
The relevant properties are:
* node was/wasn't cleaned - It's possible for a node that depends on an uncacheable node to be "cleaned", which means that our result value is `None` and skips cloning its value to complete it.
* complete as dirty/clean - This applies to both "node was/wasn't cleaned" states: a node that was cleaned without re-running uses it's previous value, but stays dirty. It's not possible for the caller of `Entry::complete` to know this though, because the previous value is stored in the Entry (without shenangians).
* (un)cacheable - When a node is uncacheable, the properties above aren't relevant; otherwise they are.

It's possible that the closure that "cleans" a Node ([here](https://github.com/pantsbuild/pants/blob/5f539c1dc6dadab4c33f451fef3103b13fd91c03/src/rust/engine/graph/src/entry.rs#L242-L261)) could clone the previous value, and then `Graph::complete` would be able to compute the final result value to pass to `Entry::complete`... but I think that would require cloning the previous result value in order to clean things, which is not cheap.

In short: I don't see a good way to do this right now.

Reply:
Perhaps it would clarify things to rename "complete_as_dirty" to "has_dirty_dependents", which makes it less verby, and clarifies that not using that value in some codepaths is "a-ok"?